source/common/stats/stats_impl.cc

```diff
@@ -152,38 +166,72 @@ TagProducerImpl::TagProducerImpl(const envoy::config::metrics::v2::StatsConfig&
   }
 }
 
-std::string TagProducerImpl::produceTags(const std::string& stat_name,
-                                         std::vector<Tag>& tags) const {
-  tags.insert(tags.end(), default_tags_.begin(), default_tags_.end());
+void TagProducerImpl::addExtractorsMatching(absl::string_view name) {
+  int num_found = 0;
+  for (const auto& desc : Config::TagNames::get().descriptorVec()) {
+    if (desc.name == name) {
+      addExtractor(Stats::TagExtractorImpl::createTagExtractor(desc.name, desc.regex));
+      ++num_found;
+    }
+  }
+  if (num_found == 0) {
+    throw EnvoyException(fmt::format(
+        "No regex specified for tag specifier and no default regex for name: '{}'", name));
+  }
+}
+
+void TagProducerImpl::addExtractor(TagExtractorPtr extractor) {
+  const absl::string_view prefix = extractor->prefixToken();
+  if (prefix.empty()) {
+    tag_extractors_without_prefix_.emplace_back(std::move(extractor));
+  } else {
+    tag_extractor_prefix_map_[prefix].emplace_back(std::move(extractor));
+  }
+}
 
+void TagProducerImpl::forEachExtractorMatching(
+    const std::string& stat_name, std::function<void(const TagExtractorPtr&)> f) const {
   IntervalSetImpl<size_t> remove_characters;
-  for (const TagExtractorPtr& tag_extractor : tag_extractors_) {
-    tag_extractor->extractTag(stat_name, tags, remove_characters);
+  for (const TagExtractorPtr& tag_extractor : tag_extractors_without_prefix_) {
+    f(tag_extractor);
+  }
+  const std::string::size_type dot = stat_name.find('.');
+  if (dot != std::string::npos) {
+    const absl::string_view token = absl::string_view(stat_name.data(), dot);
+    const auto iter = tag_extractor_prefix_map_.find(token);
+    if (iter != tag_extractor_prefix_map_.end()) {
+      for (const TagExtractorPtr& tag_extractor : iter->second) {
+        f(tag_extractor);
+      }
+    }
   }
-  return StringUtil::removeCharacters(stat_name, remove_characters);
 }
 
-// Roughly estimate the size of the vectors.
+std::string TagProducerImpl::produceTags(const std::string& metric_name,
+                                         std::vector<Tag>& tags) const {
+  tags.insert(tags.end(), default_tags_.begin(), default_tags_.end());
+  IntervalSetImpl<size_t> remove_characters;
+  forEachExtractorMatching(
+      metric_name, [&remove_characters, &tags, metric_name](const TagExtractorPtr& tag_extractor) {
```

Comment:
Good question...I was thinking it made sense to capture the reference only if it was going to be modified in the lambda body.

The first example in https://google.github.io/styleguide/cppguide.html#Lambda_expressions supports my pattern but doesn't explicitly require it.  I don't see any overriding commentary in https://github.com/envoyproxy/envoy/blob/master/STYLE.md

Reply:
Hmmmm, since `metric_name` is a const ref already, I think the only difference would be whether you make a copy or not, since the lambda should retain the constness of the ref.  IIUC, this is a similar tradeoff to passing a string by value or by const ref in a function if the function is not going to modify it.

Reply:
I agree, there should be no performance difference here.  I felt like prefixing a variable with "&" was a way for me to tell the reader "I'm going to be modifying this state in the lambda body".  Since I'm not planning on modifying it, I felt like I should omit it.

Reply:
I think there is a performance difference, here.  IIUC, without the `&`, you are making a local copy of the string for the lambda's use, which is unnecessary.

Reply:
TIL: const refs get value-ized by C++ in a lambda capture.  Right you are.  Switched to using & in the capture for metric_name.